File: libraryBase/src/main/java/com/aversyk/librarybase/utils/DisplayUtil.java
package com.aversyk.librarybase.utils;

import android.app.Activity;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.graphics.Rect;
import android.os.Build;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.WindowManager;

/**
 * 屏幕相关工具类
 * @author  Averysk
 */
public class DisplayUtil {

    /**
     * 获取设备信息
     *
     * @return
     */
    public static String getModel() {
        String model = null;
        model = Build.MODEL.replace(" ", "_");
        return model;
    }

    /**
     * 获取屏幕宽度【像素】
     *
     * @return
     */
    public static int getScreenWidth(Activity activity) {
        if (activity == null) {
            return 0;
        }
        DisplayMetrics dm = new DisplayMetrics();
        activity.getWindowManager().getDefaultDisplay().getMetrics(dm);
        return dm.widthPixels;// (int)(dm.widthPixels * dm.density +0.5f);
    }

    /**
     * 获取屏幕高度【像素】
     *
     * @return
     */
    public static int getScreenHeight(Activity activity) {
        if (activity == null) {
            return 0;
        }
        DisplayMetrics dm = new DisplayMetrics();
        activity.getWindowManager().getDefaultDisplay().getMetrics(dm);
        return dm.heightPixels;// (int)(dm.heightPixels * dm.density +0.5f);
    }

    /**
     * 获取屏幕尺寸
     * @param activity
     * @return 宽高
     */
    public static int[] getScreenSize(Activity activity){
        int screenSize[] = new int[2];
        DisplayMetrics displayMetrics = new DisplayMetrics();
        activity.getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        screenSize[0] = displayMetrics.widthPixels;
        screenSize[1] = displayMetrics.heightPixels;
        return screenSize;
    }

    /**
     * 获取屏幕宽高
     */
    public static int[] getScreenSize(Context context) {
        WindowManager manager = (WindowManager) context.getApplicationContext().getSystemService(Context.WINDOW_SERVICE);
        DisplayMetrics outMetrics = new DisplayMetrics();
        manager.getDefaultDisplay().getMetrics(outMetrics);
        return new int[]{outMetrics.widthPixels, outMetrics.heightPixels};
    }

    /**
     * 获取设备宽度
     */
    public static int getDeviceWidth(Context context) {
        DisplayMetrics dm = new DisplayMetrics();
        dm = context.getResources().getDisplayMetrics();
        // 屏幕宽（像素，如：480px）
        int screenWidth = dm.widthPixels;
        return screenWidth;

    }

    /**
     * 获取设备高度
     */
    public static int getDeviceHeight(Context context) {
        DisplayMetrics dm = new DisplayMetrics();
        dm = context.getResources().getDisplayMetrics();
        // 屏幕高（像素，如：800px）
        int screenHeight = dm.heightPixels;
        return screenHeight;

    }

    /**
     * 获取分辨率
     */
    public static String getResolution(Context context){
        String resolution = "";
        DisplayMetrics dm = context.getResources().getDisplayMetrics();
        int screenWidth = dm.widthPixels;
        int screenHeight = dm.heightPixels;
        resolution = screenWidth + "*" + screenHeight;
        return resolution;
    }

    /**
     * 获取手机大小（分辨率）
     */
    public static DisplayMetrics getScreenPix(Activity activity) {
        DisplayMetrics displaysMetrics = new DisplayMetrics();
        activity.getWindowManager().getDefaultDisplay().getMetrics(displaysMetrics);
        return displaysMetrics;
    }

    /**
     * 获得状态栏的高度
     */
    public static int getStatusHeight(Context context) {
        int statusHeight = -1;
        try {
            Class<?> clazz = Class.forName("com.android.internal.R$dimen");
            Object object = clazz.newInstance();
            int height = Integer.parseInt(clazz.getField("status_bar_height").get(object).toString());
            statusHeight = context.getResources().getDimensionPixelSize(height);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return statusHeight;
    }

    /**
     * 获取状态栏高度
     * @param activity
     * @return
     */
    public static int getStatusBarHeight(Activity activity){
        Rect rect = new Rect();
        activity.getWindow().getDecorView().getWindowVisibleDisplayFrame(rect);
        return rect.top==0 ? 60 : rect.top;
    }

    /**
     * 获取状态栏高度
     * @param context
     * @return
     */
    public static int getStatusBarHeight(Context context) {
        int statusBarHeight = 0;
        Resources res = context.getResources();
        int resourceId = res.getIdentifier("status_bar_height", "dimen", "android");
        if (resourceId > 0) {
            statusBarHeight = res.getDimensionPixelSize(resourceId);
        }
        return statusBarHeight;
    }

    public static int getActionBarHeight(Activity activity){
        TypedValue tv = new TypedValue();
        if (activity.getTheme().resolveAttribute(android.R.attr.actionBarSize, tv, true)) {
            return TypedValue.complexToDimensionPixelSize(tv.data, activity.getResources().getDisplayMetrics());
        }
        return 0;
    }

    /**
     * 获取内容区域
     * @param activity
     * @return
     */
    public static int getContextRect(Activity activity){
        //应用区域
        Rect outRect1 = new Rect();
        activity.getWindow().getDecorView().getWindowVisibleDisplayFrame(outRect1);
        return outRect1.height();
    }

    /**
     * 将px值转换为dip或dp值，保证尺寸大小不变
     */
    public static int px2dip(Context context, float pxValue) {
        if (context == null) {
            return 0;
        }
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (pxValue / scale + 0.5f);
    }

    /**
     * 将dip或dp值转换为px值，保证尺寸大小不变
     */
    public static int dip2px(Context context, float dipValue) {
        if (context == null) {
            return 0;
        }
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dipValue * scale + 0.5f);
    }

    /**
     * 将dp值转换为px值，保证尺寸大小不变
     */
    public static int dp2Px(Context context, float dp) {
        if (context == null) {
            return 0;
        }
        Resources r = context.getResources();
        float px = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, r.getDisplayMetrics());
        return (int) px;
    }

    /**
     * 将sp值转换为px值，保证文字大小不变
     */
    public static int sp2px(Context context, float spValue) {
        if (context == null) {
            return 0;
        }
        final float fontScale = context.getResources().getDisplayMetrics().scaledDensity;
        return (int) (spValue * fontScale + 0.5f);
    }

    /**
     * 将px值转换为sp值，保证文字大小不变
     */
    public static int px2sp(Context context, float pxValue) {
        if (context == null) {
            return 0;
        }
        final float fontScale = context.getResources().getDisplayMetrics().scaledDensity;
        return (int) (pxValue / fontScale + 0.5f);
    }

    /**
     * 获取MetaData信息
     *
     * @param name
     * @param def
     * @return
     */
    public static String getMetaDataValue(Context context, String name,
                                          String def) {
        String value = getMetaDataValue(context, name);
        return (value == null) ? def : value;
    }

    public static String getMetaDataValue(Context context, String name) {
        Object value = null;
        PackageManager packageManager = context.getPackageManager();
        ApplicationInfo applicationInfo;
        try {
            applicationInfo = packageManager.getApplicationInfo(
                    context.getPackageName(), PackageManager.GET_META_DATA);
            if (applicationInfo != null && applicationInfo.metaData != null) {
                value = applicationInfo.metaData.get(name);
            }
        } catch (PackageManager.NameNotFoundException e) {
            throw new RuntimeException(
                    "Could not read the name in the manifest file.", e);
        }
        if (value == null) {
            throw new RuntimeException("The name '" + name
                    + "' is not defined in the manifest file's meta data.");
        }
        return value.toString();
    }
}

File: libraryBase/src/main/java/com/aversyk/librarybase/utils/DisplayUtil.kt
package com.aversyk.librarybase.utils

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.graphics.Rect
import android.os.Build
import android.util.DisplayMetrics
import android.util.TypedValue
import android.view.WindowManager
import com.aversyk.librarybase.R

/**
 * 屏幕相关工具类
 * @author  Averysk
 */
object DisplayUtil {

    /**
     * 获取设备信息
     *
     * @return
     */
    val model: String
        get() {
            return Build.MODEL.replace(" ", "_")
        }

    /**
     * 获取屏幕宽度【像素】
     *
     * @return
     */
    fun getScreenWidth(activity: Activity?): Int {
        if (activity == null) {
            return 0
        }
        val dm = DisplayMetrics()
        activity.windowManager.defaultDisplay.getMetrics(dm)
        return dm.widthPixels // (int)(dm.widthPixels * dm.density +0.5f);
    }

    /**
     * 获取屏幕高度【像素】
     *
     * @return
     */
    fun getScreenHeight(activity: Activity?): Int {
        if (activity == null) {
            return 0
        }
        val dm = DisplayMetrics()
        activity.windowManager.defaultDisplay.getMetrics(dm)
        return dm.heightPixels // (int)(dm.heightPixels * dm.density +0.5f);
    }

    /**
     * 获取屏幕尺寸
     * @param activity
     * @return 宽高
     */
    fun getScreenSize(activity: Activity): IntArray {
        val screenSize = IntArray(2)
        val displayMetrics = DisplayMetrics()
        activity.windowManager.defaultDisplay.getMetrics(displayMetrics)
        screenSize[0] = displayMetrics.widthPixels
        screenSize[1] = displayMetrics.heightPixels
        return screenSize
    }

    /**
     * 获取屏幕宽高
     */
    fun getScreenSize(context: Context): IntArray {
        val manager = context.applicationContext.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val outMetrics = DisplayMetrics()
        manager.defaultDisplay.getMetrics(outMetrics)
        return intArrayOf(outMetrics.widthPixels, outMetrics.heightPixels)
    }

    /**
     * 获取设备宽度
     */
    fun getDeviceWidth(context: Context): Int {
        val dm: DisplayMetrics = context.resources.displayMetrics
        // 屏幕宽（像素，如：480px）
        return dm.widthPixels
    }

    /**
     * 获取设备高度
     */
    fun getDeviceHeight(context: Context): Int {
        val dm: DisplayMetrics = context.resources.displayMetrics
        // 屏幕高（像素，如：800px）
        return dm.heightPixels
    }

    /**
     * 获取分辨率
     */
    fun getResolution(context: Context): String {
        val dm = context.resources.displayMetrics
        val screenWidth = dm.widthPixels
        val screenHeight = dm.heightPixels
        return "$screenWidth*$screenHeight"
    }

    /**
     * 获取手机大小（分辨率）
     */
    fun getScreenPix(activity: Activity): DisplayMetrics {
        val displaysMetrics = DisplayMetrics()
        activity.windowManager.defaultDisplay.getMetrics(displaysMetrics)
        return displaysMetrics
    }

    /**
     * 获得状态栏的高度
     */
    @SuppressLint("PrivateApi")
    fun getStatusHeight(context: Context): Int {
        var statusHeight = -1
        try {
            val clazz = Class.forName("com.android.internal.R\$dimen")
            val `object` = clazz.newInstance()
            val height = clazz.getField("status_bar_height")[`object`].toString().toInt()
            statusHeight = context.resources.getDimensionPixelSize(height)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return statusHeight
    }

    /**
     * 获取状态栏高度
     * @param activity
     * @return
     */
    fun getStatusBarHeight(activity: Activity): Int {
        val rect = Rect()
        activity.window.decorView.getWindowVisibleDisplayFrame(rect)
        return if (rect.top == 0) 60 else rect.top
    }

    /**
     * 获取状态栏高度
     * @param context
     * @return
     */
    fun getStatusBarHeight(context: Context): Int {
        var statusBarHeight = 0
        val res = context.resources
        val resourceId = res.getIdentifier("status_bar_height", "dimen", "android")
        if (resourceId > 0) {
            statusBarHeight = res.getDimensionPixelSize(resourceId)
        }
        return statusBarHeight
    }

    /**
     * 获取状态栏高度
     * @param activity
     * @return
     */
    fun getActionBarHeight(activity: Activity): Int {
        val tv = TypedValue()
        return if (activity.theme.resolveAttribute(R.attr.actionBarSize, tv, true)) {
            TypedValue.complexToDimensionPixelSize(tv.data, activity.resources.displayMetrics)
        } else 0
    }

    /**
     * 获取内容区域
     * @param activity
     * @return
     */
    fun getContextRect(activity: Activity): Int {
        //应用区域
        val outRect1 = Rect()
        activity.window.decorView.getWindowVisibleDisplayFrame(outRect1)
        return outRect1.height()
    }

    /**
     * 将px值转换为dip或dp值，保证尺寸大小不变
     */
    fun px2dip(context: Context?, pxValue: Float): Int {
        if (context == null) {
            return 0
        }
        val scale = context.resources.displayMetrics.density
        return (pxValue / scale + 0.5f).toInt()
    }

    /**
     * 将dip或dp值转换为px值，保证尺寸大小不变
     */
    fun dip2px(context: Context?, dipValue: Float): Int {
        if (context == null) {
            return 0
        }
        val scale = context.resources.displayMetrics.density
        return (dipValue * scale + 0.5f).toInt()
    }

    /**
     * 将dp值转换为px值，保证尺寸大小不变
     */
    fun dp2Px(context: Context?, dp: Float): Int {
        if (context == null) {
            return 0
        }
        val r = context.resources
        val px = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, r.displayMetrics)
        return px.toInt()
    }

    /**
     * 将sp值转换为px值，保证文字大小不变
     */
    fun sp2px(context: Context?, spValue: Float): Int {
        if (context == null) {
            return 0
        }
        val fontScale = context.resources.displayMetrics.scaledDensity
        return (spValue * fontScale + 0.5f).toInt()
    }

    /**
     * 将px值转换为sp值，保证文字大小不变
     */
    fun px2sp(context: Context?, pxValue: Float): Int {
        if (context == null) {
            return 0
        }
        val fontScale = context.resources.displayMetrics.scaledDensity
        return (pxValue / fontScale + 0.5f).toInt()
    }

    /**
     * 获取MetaData信息
     *
     * @param name
     * @param def
     * @return
     */
    fun getMetaDataValue(context: Context, name: String, def: String): String {
        return getMetaDataValue(context, name)
    }

    fun getMetaDataValue(context: Context, name: String): String {
        var value: Any? = null
        val packageManager = context.packageManager
        val applicationInfo: ApplicationInfo?
        try {
            applicationInfo = packageManager.getApplicationInfo(context.packageName, PackageManager.GET_META_DATA)
            if (applicationInfo?.metaData != null) {
                value = applicationInfo.metaData[name]
            }
        } catch (e: PackageManager.NameNotFoundException) {
            throw RuntimeException("Could not read the name in the manifest file.", e)
        }
        if (value == null) {
            throw RuntimeException("The name '$name' is not defined in the manifest file's meta data.")
        }
        return value.toString()
    }
}